### ISSUE TITLE
Chapel generates invalid JSON for lists and maps containing records

### ISSUE THREAD

**Issue by hokiegeek2** [chapel-lang/chapel#18156]
### Summary of Problem
The Chapel Json generation works really well for a variety of use cases involving records-> well done! The only issues I've seen are JSON generation of lists and maps containing records 

### Steps to Reproduce

I wrote a simple Chapel app named json_to_object_test.chpl

```
module JsonStuff {
    use IO;
    use List;
    use Map;

    proc main() {
        enum FooType { GOOD,BAD,UGLY }
        record Foo {

            var name: string;
            var id: int;
            var isFoo: bool;
            var fooType: FooType;
        }

        record Bar {
            var name: string;
            var id: int;
            var isBar: bool;
        }

        record FooBar {
            var name: string;
            var id: int;
            var isFooBar : bool;
            var foo: Foo;
            var bar: Bar;
            proc init(name: string, id: int, isFooBar : bool, foo: Foo, bar: Bar) {
                this.name=name;
                this.id=id;
                this.isFooBar = isFooBar;
                this.foo = foo;
                this.bar = bar;
            }
            proc init() {}
        }

        record FooBarGroup {
            var name: string;
            var id: int;
            var isFooBar : bool;
            var foos: list;
            var bars: list;
            proc init(name: string, id: int, isFooBar : bool, foos,  bars) {
                this.name=name;
                this.id=id;
                this.isFooBar = isFooBar;
                this.foos = foos;
                this.bars = bars;
            }
            proc init() {}
        }

        var foo = new Foo('fooOne', 1, true, FooType.BAD);
        var fooTwo = new Foo('fooTwo', 2, false, FooType.GOOD);
        var bar = new Bar('barOne', 1, false);
        var barTwo = new Bar('barTwo', 2, true);
        var fb = new FooBar('fbOne', 2, false, foo, bar);
        var fbTwo = new FooBar('fbTwo', 2, false, foo, bar);
        try! writeln("Writing FooBar to JSON:  %jt".format(fb));
        var mem = try! openmem();
        var writer = try! mem.writer().write("%jt".format(fb));
        var reader = try! mem.reader();

        var f = new FooBar();
        try! reader.readf("%jt", f);

        try! writeln("Writing FooBar array to a JSON array %jt".format([fb,fbTwo]));
        try! writeln("Writing FooBar list to a JSON array %jt".format(new list([fb,fbTwo])));

        var fbg = new FooBarGroup('groupOne', 1, true, new list([foo,fooTwo]),new list([bar,barTwo]));
        var fbgJson = try! "%jt".format(fbg);
        try! writeln("Writing FooBarGroup to JSON: %jt".format(fbg));

        var fbm = new map(keyType=string,valType=FooBar);
        fbm.add('one', fb);
        fbm.add('two', fbTwo);

        try! writeln("Writing FooBar map to JSON: %jt".format(fbm));
    }
}
```

**Compile command:**
chpl json_to_object_test.chpl

**Execution command:**
././json_to_object_test -nl 1

**Results**
Writing FooBar to JSON:  {"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}
Writing FooBar array to a JSON array [{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}, {"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}]
Writing FooBar list to a JSON array "["{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}", "{"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}"]"
Writing FooBarGroup to JSON: {"name":"groupOne", "id":1, "isFooBar":true, "foos":"["{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}", "{"name":"fooTwo", "id":2, "isFoo":false, "fooType":"GOOD"}"]", "bars":"["{"name":"barOne", "id":1, "isBar":false}", "{"name":"barTwo", "id":2, "isBar":true}"]"}
Writing FooBar map to JSON: "{""two"": "{"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}", ""one"": "{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}"}"

The JSON strings for both the FooBar list and string -> FooBar map are incorrect

FooBar list is:

```
"["{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}", "{"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}"]"
```

In python:

```
>>> json.loads('"["{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}", "{"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}"]"')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kjyost/miniconda3/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/Users/kjyost/miniconda3/lib/python3.7/json/decoder.py", line 340, in decode
    raise JSONDecodeError("Extra data", s, end)
json.decoder.JSONDecodeError: Extra data: line 1 column 4 (char 3)
```

FooBar list should be:
```
[{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}, {"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}]
````

In python:
```
>>> json.loads('[{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}, {"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}]')
[{'name': 'fbOne', 'id': 2, 'isFooBar': False, 'foo': {'name': 'fooOne', 'id': 1, 'isFoo': True, 'fooType': 'BAD'}, 'bar': {'name': 'barOne', 'id': 1, 'isBar': False}}, {'name': 'fbTwo', 'id': 2, 'isFooBar': False, 'foo': {'name': 'fooOne', 'id': 1, 'isFoo': True, 'fooType': 'BAD'}, 'bar': {'name': 'barOne', 'id': 1, 'isBar': False}}]
>>> 
```

FooBar map is:

```
"{""two"": "{"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}", ""one"": "{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}"}"
```

In python:

```
>>> json.loads('"{""two"": "{"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}", ""one"": "{"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}"}"')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kjyost/miniconda3/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/Users/kjyost/miniconda3/lib/python3.7/json/decoder.py", line 340, in decode
    raise JSONDecodeError("Extra data", s, end)
json.decoder.JSONDecodeError: Extra data: line 1 column 4 (char 3)
>>> 
```

FooBar map should be:

```
{"two": {"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}, "one": {"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}}
```

In python:

```
>>> json.loads('{"two": {"name":"fbTwo", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}, "one": {"name":"fbOne", "id":2, "isFooBar":false, "foo":{"name":"fooOne", "id":1, "isFoo":true, "fooType":"BAD"}, "bar":{"name":"barOne", "id":1, "isBar":false}}}')
{'two': {'name': 'fbTwo', 'id': 2, 'isFooBar': False, 'foo': {'name': 'fooOne', 'id': 1, 'isFoo': True, 'fooType': 'BAD'}, 'bar': {'name': 'barOne', 'id': 1, 'isBar': False}}, 'one': {'name': 'fbOne', 'id': 2, 'isFooBar': False, 'foo': {'name': 'fooOne', 'id': 1, 'isFoo': True, 'fooType': 'BAD'}, 'bar': {'name': 'barOne', 'id': 1, 'isBar': False}}}
>>> 

```
### Configuration Information

- Output of `chpl --version`: 1.24.1

**Comment by bradcray** [comment]
@hokiegeek2 : Apologies for the lack of response on this.  Our "new issue bot" went AWOL this week so I didn't notice it until now (and suspect others may be in the same camp).  I'm tagging @mppf on this as the person who's done the most with JSON support in Chapel and @dlongnecke-cray and @daviditen as primary contributors to list and map.

**Comment by hokiegeek2** [comment]
@bradcray No worries! Like I wrote in the issue, the current JSON support is awesome and fixing these two issues will make it even awesomer. 👍

**Comment by mppf** [comment]
I have a PR that should fix this in #18167. @hokiegeek2 this should allow JSON *output* for list and map but JSON *input* for these types still does not work. Please let us know if that presents a problem for you.

**Comment by hokiegeek2** [comment]
@mppf Awesome! Thanks a bunch for working this so quickly!

Object -> JSON via Chapel is key for us as JSON -> Object is handled on the Python side, so JSON output is all we need.

This is awesome, thanks again!

@glitch FYI
